web/modules/custom/dpl_related_content/src/Services/RelatedContent.php

```diff
@@ -0,0 +1,371 @@
+      $node_ids = $this->getNodeIds($excluded_uuid, $tags, $categories);
+      $event_ids = $this->getEventInstanceIds($excluded_uuid, $tags, $categories);
+    }
+
+    // If we found less than minimum results, we'll add branches to the mix in
+    // addition to tags and categories.
+    if ((count($event_ids) + count($node_ids) < $this->minItems) && !empty($branches)) {
+      $node_ids = $this->getNodeIds($excluded_uuid, $tags, $categories, $branches);
+      $event_ids = $this->getEventInstanceIds($excluded_uuid, $tags, $categories, $branches);
+    }
+
+    // If the count is still under minimum, we'll find the upcoming events,
+    // and the latest nodes instead.
+    if (count($event_ids) + count($node_ids) < $this->minItems) {
+      $node_ids = $this->getNodeIds($excluded_uuid);
+      $event_ids = $this->getEventInstanceIds($excluded_uuid);
+    }
+
+    // If we still have less than minimum, we just won't display anything.
+    if (count($event_ids) + count($node_ids) < $this->minItems) {
+      return [];
+    }
+
+    $events = $this->entityTypeManager->getStorage('eventinstance')->loadMultiple($event_ids);
+    $nodes = $this->entityTypeManager->getStorage('node')->loadMultiple($node_ids);
+
+    // We want to combine events and nodes into a single array, however,
+    // we want to interlace them, so it becomes mixed, rather than just
+    // all the nodes and then all the events.
+    // We could also sort the merging based on the actual date values, but this
+    // will be too expensive for what it is worth.
+    $content = [];
+
+    // Remove array keys, so we can do the sorting.
+    $events = array_values($events);
+    $nodes = array_values($nodes);
+
+    $node_view_builder = $this->entityTypeManager->getViewBuilder('node');
+    $event_view_builder = $this->entityTypeManager->getViewBuilder('eventinstance');
+
+    // We find the longer array, to do a simple FOR loop.
+    $length = max(count($events), count($nodes));
+
+    for ($i = 0; $i < $length; $i++) {
+      if ($i >= $this->maxItems) {
+        break;
+      }
+
+      if (isset($events[$i])) {
+        $content[] = $event_view_builder->view($events[$i], 'card');
+      }
+
+      if (isset($nodes[$i])) {
+        $content[] = $node_view_builder->view($nodes[$i], 'card');
+      }
+    }
+
+    return [
+      '#theme' => 'dpl_related_content_slider',
+      '#items' => $content,
+      // The results should be cached, but, they have a lot of dependencies -
+      // even depending on the current time (finding future events).
+      // The individual peices of content are cached themselves, so for the full
+      // result list, we'll add an easily-invalidated cache, with a bunch of
+      // cache tags, and even a time-based cache.
+      '#cache' => [
+        // Max age of 12 hours.
+        '#max-age' => (60 * 60 * 12),
+        '#tags' => ['eventinstance_list', 'eventseries_list', 'node_list'],
```

Comment:
:+1: The cache tags here makes sense.

The max age is set to 12 hours. This might be good and it might not.

As the values here depend on the date I suggest you instead add a cache context with the date. This will automatically invalidate entries when the date changes.

Reply:
@kasperg 
>I suggest you instead add a cache context with the date

Isnt that the same as `max-age`? What context do you have in mind?